Create IDP best practices section and four factors conceptual doc (#15479)
pulumi/docs: content/docs/idp/best-practices/four-factors.md

---
title: "Four Factors: Templates, Components, Environments, and Policies"
linktitle: "The Four Factors Framework"
meta_desc: A framework for building with Pulumi IDP using Templates, Components, Environments, and Policies
allow_long_title: true
menu:
  idp:
    parent: idp-best-practices
    weight: 10
---

This guide introduces a framework for building with Pulumi IDP that leverages four key factors to create secure, scalable, and flexible developer workflows. By understanding and effectively combining these factors, platform teams can build robust internal developer platforms that enable developers to provision infrastructure efficiently while maintaining compliance and security.

## Templates

> Templates are user-configurable scaffolding that creates new instances of infrastructure resources. They serve as the entry point for developers to quickly bootstrap new projects or components with pre-configured settings and best practices baked in.

Author: ```suggestion
[Templates](/docs/pulumi-cloud/developer-platforms/templates/) are user-configurable scaffolding that creates new instances of infrastructure resources. They serve as the entry point for developers to quickly bootstrap new projects or components with pre-configured settings and best practices baked in.
```

Key characteristics:

- **User-configurable**: Developers can customize templates through simple parameters

> - **Instance creation**: Each template use creates a new, independent [Pulumi stack](/docs/iac/concepts/stacks/)

Reviewer: Is `Instance creation` what was intended? If so, I'm not sure what that means

Author: Good call out. The language there was a little squirrelly. Fixed it.

> - **Instance creation**: Each template use creates a new, independent [Pulumi stack](/docs/iac/concepts/stacks/)

Author: ```suggestion
- **Stack creation**: Each time a template is used it creates one or more new, independent [Pulumi stacks](/docs/iac/concepts/stacks/)
```

- **Standardized starting point**: Ensures consistent project structure and configuration
- **Low barrier to entry**: Enables developers to get started quickly without deep infrastructure knowledge

Templates abstract away the complexity of initial setup while providing flexibility for customization based on specific use cases.

## Components

> Components are reusable building blocks that abstract and encapsulate business logic, set sensible defaults, limit inputs, and encode company standards. They represent the core infrastructure primitives that your organization uses repeatedly.

Reviewer: Above, `stack` is a link. In the rest of the page, we don't link to our core constructs (component, environment, policy) -- should we?

Key characteristics:

- **Business logic abstraction**: Hide complex infrastructure patterns behind simple interfaces
- **Sensible defaults**: Provide secure, compliant defaults out of the box
- **Limited inputs**: Reduce cognitive load by exposing only necessary configuration options
- **Company standards**: Encode organizational policies and best practices directly in code

Components enable platform teams to create opinionated infrastructure building blocks that developers can use confidently, knowing they follow organizational standards.

## Environments

> Environments provide just enough credentials and configuration for your needs through composable, secure configuration management. They ensure that the right secrets and settings are available to the right workloads at the right time.

Reviewer: ```suggestion
Environments provide credentials and configuration for your needs through composable, secure configuration management. They ensure that the right secrets and settings are available to the right workloads at the right time.
```
